kingpin/actors/aws/api_call_queue.py

```diff
@@ -0,0 +1,109 @@
+from boto import exception as boto_exception
+from botocore import exceptions as botocore_exceptions
+from tornado import concurrent
+from tornado import gen
+from tornado import queues
+from tornado import ioloop
+
+EXECUTOR = concurrent.futures.ThreadPoolExecutor(10)
+
+
+class ApiCallQueue:
+    """
+    Handles queueing up and sending AWS api calls serially,
+    with exponential backoff when there is throttling.
+
+    Supports both boto2 and boto3.
+    """
+
+    def __init__(self):
+        self.executor = EXECUTOR
+
+        self._queue = queues.Queue()
+        ioloop.IOLoop.current().spawn_callback(self._process_queue)
+
+        # Used for controlling how fast the work queue is processed,
+        # with exponential delay on throttling errors.
+        self.delay_min = 0.25
+        self.delay_max = 30
+        # We don't have a delay until we first get throttled.
+        self.delay = 0
+
+        # There are a number of different rate limiting messages
+        # boto2 can return when rate limits are reached, depending
+        # on which apis are used.
+        self.boto2_throttle_strings = (
+            'Throttling',
+            'Rate exceeded',
+            'reached max retries',
+        )
+
+    @gen.coroutine
+    def call(self, api_function, *args, **kwargs):
+        result_queue = queues.Queue(maxsize=1)
+        yield self._queue.put((result_queue, api_function, args, kwargs))
+        result = yield result_queue.get()
```

Comment:
Is there a timeout here on this at all?

Reply:
No, and that's by design.

If an API call is being rate limited, this will back off the attempts and try forever until it stops getting rate limited. The API calls themselves will have some timeout from boto, which'll cause an exception to raise, and the Actors have their own run timeout that wraps this.

It may make sense to add something like a 5 minute timeout here, in case we are getting endlessly rate limited, but why? I feel like trying continuously with backoff for the rate limit case is a fine way to handle rate limiting.

It could use some logging when it does get rate limited, though.